src/crimson/osd/backfill_state.cc

```diff
@@ -392,16 +392,25 @@ BackfillState::Enqueuing::Enqueuing(my_context ctx)
     }
   } while (!all_emptied(primary_bi, backfill_state().peer_backfill_info));
 
-  if (backfill_state().progress_tracker->tracked_objects_completed()
-      && Enqueuing::all_enqueued(peering_state(),
-				 backfill_state().backfill_info,
-				 backfill_state().peer_backfill_info)) {
-    backfill_state().last_backfill_started = hobject_t::get_max();
-    backfill_listener().update_peers_last_backfill(hobject_t::get_max());
+  if (should_rescan_primary(backfill_state().peer_backfill_info,
+				   primary_bi)) {
+    // need to grab one another chunk of the object namespace and restart
+    // the queueing.
+    logger().debug("reached end for current local chunk", __func__);
+    post_event(RequestPrimaryScanning{});
+    return;
+  } else {
```

Comment:
I'm not entirely sure I understand why we might need an additional primary re-scan in the issue attached. The PG enters backfill and scans the primary, the scan resulted in an identical replica scans which makes the backfill to keep waiting (endlessly). Why are we expecting the rescan to result differently?

Reply:
Rescanning the primary will discover new objects to push, if there are objects not yet backfilled, the backfill will go on; if not, eventually, all objects are scanned and the backfill finishes. Am I right?

Reply:
Because no push is enqueued, the loop above must have exhausted all objects scanned by the primary, so either the primary needs to be rescanned or the backfill is finished. Am I right?

Reply:
This sounds reasonable to me